File: Stack/20. Valid Parentheses/validParentheses.js
/**
 * Time: O(n) | Space: O(n)
 * @param {string} s
 * @return {boolean}
 */
const isValid = function(s) {
  const map = {
    ")": "(",
    "]": "[",
    "}": "{"
  };
  const stack = [];

  for (const c of s) {
    if (!(c in map)) {
      stack.push(c);
      continue;
    }
    if (stack.length === 0 || map[c] !== stack[stack.length-1]) {
      return false;
    }
    stack.pop();
  }

  return stack.length === 0;
};

File: Stack/20. Valid Parentheses/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        hashmap = {
            ")": "(",
            "]": "[",
            "}": "{"
        }

        for c in s:
            if c not in hashmap:
                stack.append(c)
                continue
            if not stack or hashmap[c] != stack[-1]:
                return False
            stack.pop()
        
        return not stack

